File: Data Structures & Algorithms/interleaving-string/submission-3.cpp
class Solution {
    int x[101][101]{};
public:
    bool isInterleave(string s1, string s2, string s3) {
        auto dp = [&](auto&& dp, int i, int j, int t) {
            if (t == s3.size() && i == s1.size() && s2.size() == j) return true;
            bool res = false;
            if (x[i][j] == 1) return false;
            if (x[i][j] == 2) return true; 
            if (s1[i] == s3[t]){
                res = res || dp(dp, i + 1, j, t + 1);
            }
            if (s2[j] == s3[t]) {
                res = res || dp(dp, i, j + 1, t + 1);
            }
            if (res) x[i][j] = 2;
            else x[i][j] = 1;
            return res;
        };
        return dp(dp, 0, 0, 0);
    }
};

File: Data Structures & Algorithms/majority-element-ii/submission-3.py
class Solution:
    def majorityElement(self, nums: List[int]) -> List[int]:
        cand1 = nums[0]
        cand2 = float("-inf")
        cnt1 = 1
        cnt2 = 0
        for i in range(len(nums)):
            num = nums[i]
            if num == cand1:
                cnt1 += 1
            elif num == cnt2:
                cnt2 += 1
            else :
                cnt2 -= 1
            if cnt2 < 0 :
                cand2 = num
                cnt2 = 1
            if cnt2 > cnt1:
                cnt1, cnt2 = cnt2, cnt1
                cand1, cand2 = cand2, cand1
        cnt1 = 0
        cnt2 = 0
        res = []
        for num in nums:
            if num == cand1:
                cnt1 += 1
            if num == cand2:
                cnt2 += 1
        if cnt1 >  len(nums) //3:
            res.append(cand1)
        if cnt2 > len(nums) // 3:
            res.append(cand2)
        return res    
